File: IntRuoyiBackend/yudao-module-dcc/src/test/java/cn/iocoder/yudao/module/dcc/DccControlledPrintContractTest.java
package cn.iocoder.yudao.module.dcc;

import org.junit.jupiter.api.Test;

import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

class DccControlledPrintContractTest {

    private static final Path BACKEND_ROOT = resolveBackendRoot(Path.of("").toAbsolutePath());
    private static final Path REPO_ROOT = BACKEND_ROOT.getParent();

    @Test
    void schemaDefinesControlledPrintRecordTable() throws Exception {
        String baseSchema = readBackend("sql/mysql/20260513_dcc_base_schema.sql");
        String testSchema = readBackend("yudao-module-dcc/src/test/resources/sql/create_tables.sql");
        assertControlledPrintSchema(baseSchema, "base schema");
        assertControlledPrintSchema(testSchema, "test schema");
    }

    @Test
    void backendExposesFormalControlledPrintApiAndPermission() throws Exception {
        String controller = readBackend("yudao-module-dcc/src/main/java/cn/iocoder/yudao/module/dcc/controller/admin/file/DccControlledFileController.java");
        assertTrue(controller.contains("DccControlledFilePrintService"),
                "controller must use a formal controlled print service");
        assertTrue(controller.contains("dcc:controlled-file:print"),
                "controlled print write endpoints must require dcc:controlled-file:print");
        assertTrue(controller.contains("/{id:\\\\d+}/controlled-print"),
                "controller must expose controlled print creation endpoint under the controlled file");
        assertTrue(controller.contains("/{id:\\\\d+}/controlled-print/records"),
                "controller must expose controlled print records endpoint under the controlled file");
        assertTrue(controller.contains("controlled-print/print-html"),
                "controller must expose generated controlled print HTML with watermarked metadata");

        assertTrue(Files.exists(BACKEND_ROOT.resolve("yudao-module-dcc/src/main/java/cn/iocoder/yudao/module/dcc/service/file/DccControlledFilePrintService.java")),
                "controlled print service interface must exist");
        assertTrue(Files.exists(BACKEND_ROOT.resolve("yudao-module-dcc/src/main/java/cn/iocoder/yudao/module/dcc/dal/dataobject/file/DccControlledFilePrintRecordDO.java")),
                "controlled print record DO must exist");
        assertTrue(Files.exists(BACKEND_ROOT.resolve("yudao-module-dcc/src/main/java/cn/iocoder/yudao/module/dcc/dal/mysql/file/DccControlledFilePrintRecordMapper.java")),
                "controlled print record mapper must exist");
    }

    @Test
    void backendPrintHtmlExposesDirectPrintPolicyAndCopyNumbers() throws Exception {
        String service = readBackend("yudao-module-dcc/src/main/java/cn/iocoder/yudao/module/dcc/service/file/DccControlledFilePrintServiceImpl.java");
        assertTrue(service.contains("副本编号"),
                "generated controlled print HTML must expose per-copy copy numbers");
        assertTrue(service.contains("buildControlledPrintCopyNumbers"),
                "copy numbers must be derived consistently from print number and copies");
        assertTrue(service.contains("直接受控打印"),
                "generated controlled print HTML must expose the current direct print/no approval policy");
    }

    @Test
    void backendPrintEligibilityDoesNotRequireGeneratedArtifactIds() throws Exception {
        String forbiddenArtifactGate = "getPublishedFileId() == null && file.getStampedFileId() == null";
        String printService = readBackend("yudao-module-dcc/src/main/java/cn/iocoder/yudao/module/dcc/service/file/DccControlledFilePrintServiceImpl.java");
        String queryService = readBackend("yudao-module-dcc/src/main/java/cn/iocoder/yudao/module/dcc/service/file/DccControlledFileQueryServiceImpl.java");
        assertFalse(printService.contains(forbiddenArtifactGate),
                "controlled print creation must not reject the current ACTIVE file only because generated artifact ids are empty");
        assertFalse(queryService.contains(forbiddenArtifactGate),
                "action projection must expose PRINT for current ACTIVE files with PRINT permission even if generated artifact ids are empty");
    }

    @Test
    void backendDefinesFailFastControlledPrintErrors() throws Exception {
        String errors = readBackend("yudao-module-dcc/src/main/java/cn/iocoder/yudao/module/dcc/enums/ErrorCodeConstants.java");
        assertTrue(errors.contains("CONTROLLED_FILE_PRINT_NOT_ALLOWED"),
                "missing permission/current-version controlled print error");
        assertTrue(errors.contains("CONTROLLED_FILE_PRINT_REQUIRED_FIELD_MISSING"),
                "missing required-field controlled print error");
    }

    private static void assertControlledPrintSchema(String schema, String label) {
        assertTrue(schema.contains("CREATE TABLE IF NOT EXISTS `dcc_controlled_file_print_record`"),
                label + " must create dcc_controlled_file_print_record");
        assertTrue(schema.contains("`controlled_file_id` BIGINT NOT NULL"),
                label + " must persist controlled_file_id");
        assertTrue(schema.contains("`print_no` VARCHAR(64) NOT NULL"),
                label + " must persist a traceable print_no");
        assertTrue(schema.contains("`purpose` VARCHAR(255) NOT NULL"),
                label + " must persist print purpose");
        assertTrue(schema.contains("`copies` INT NOT NULL"),
                label + " must persist print copies");
        assertTrue(schema.contains("`receiving_department` VARCHAR(128) NOT NULL"),
                label + " must persist receiving department");
        assertTrue(schema.contains("`use_location` VARCHAR(128) NOT NULL"),
                label + " must persist use location");
        assertTrue(schema.contains("`print_user_id` BIGINT NOT NULL"),
                label + " must persist print user");
        assertTrue(schema.contains("`print_time` DATETIME NOT NULL"),
                label + " must persist print time");
        assertTrue(schema.contains("`approval_status` VARCHAR(32) NOT NULL"),
                label + " must persist direct/approval status");
    }

    private static String readBackend(String relativePath) throws Exception {
        return Files.readString(BACKEND_ROOT.resolve(relativePath), StandardCharsets.UTF_8);
    }

    private static Path resolveBackendRoot(Path start) {
        Path current = start;
        while (current != null) {
            if (Files.exists(current.resolve("sql/mysql/20260513_dcc_base_schema.sql"))
                    && Files.exists(current.resolve("yudao-module-dcc/src/main/java/cn/iocoder/yudao/module/dcc/controller/admin/file/DccControlledFileController.java"))) {
                return current;
            }
            current = current.getParent();
        }
        throw new IllegalStateException("Cannot resolve IntRuoyi backend root from " + start);
    }

    @SuppressWarnings("unused")
    private static String readRepo(String relativePath) throws Exception {
        return Files.readString(REPO_ROOT.resolve(relativePath), StandardCharsets.UTF_8);
    }
}

File: IntRuoyiBackend/script/tests/dcc_controlled_print_contract.test.mjs
import assert from 'node:assert/strict'
import fs from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'

const backendRoot = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..', '..')

const readSource = (relativePath) => {
  const absolutePath = path.join(backendRoot, relativePath)
  assert.equal(fs.existsSync(absolutePath), true, `missing required file: ${relativePath}`)
  return fs.readFileSync(absolutePath, 'utf8')
}

const assertControlledPrintSchema = (schema, label) => {
  assert.match(
    schema,
    /CREATE TABLE IF NOT EXISTS `dcc_controlled_file_print_record`/,
    `${label} must create dcc_controlled_file_print_record`
  )
  for (const column of [
    '`controlled_file_id` BIGINT NOT NULL',
    '`print_no` VARCHAR(64) NOT NULL',
    '`purpose` VARCHAR(255) NOT NULL',
    '`copies` INT NOT NULL',
    '`receiving_department` VARCHAR(128) NOT NULL',
    '`use_location` VARCHAR(128) NOT NULL',
    '`print_user_id` BIGINT NOT NULL',
    '`print_time` DATETIME NOT NULL',
    '`approval_status` VARCHAR(32) NOT NULL'
  ]) {
    assert.match(schema, new RegExp(column.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')), `${label} missing ${column}`)
  }
}

assertControlledPrintSchema(readSource('sql/mysql/20260513_dcc_base_schema.sql'), 'base schema')
assertControlledPrintSchema(
  readSource('yudao-module-dcc/src/test/resources/sql/create_tables.sql'),
  'test schema'
)

const controlledPrintMigration = readSource('sql/mysql/20260802_dcc_controlled_print_record.sql')
assert.match(
  controlledPrintMigration,
  /@dcc_controlled_print_preferred_menu_id_blocked/,
  'controlled print migration must detect occupied preferred menu id'
)
assert.match(
  controlledPrintMigration,
  /COALESCE\(MAX\(`id`\), 0\) \+ 1 FROM `system_menu`/,
  'controlled print migration must allocate a safe menu id when the preferred id is occupied'
)
assert.doesNotMatch(
  controlledPrintMigration,
  /SELECT\s+6813,\s*[\s\S]{0,200}'dcc:controlled-file:print'/,
  'controlled print migration must not rely on occupied legacy menu id 6813'
)

const controller = readSource(
  'yudao-module-dcc/src/main/java/cn/iocoder/yudao/module/dcc/controller/admin/file/DccControlledFileController.java'
)
assert.match(controller, /DccControlledFilePrintService/, 'controller must use DccControlledFilePrintService')
assert.match(controller, /dcc:controlled-file:print/, 'controlled print endpoints must require print permission')
assert.match(controller, /\/\{id:\\\\d\+\}\/controlled-print/, 'missing controlled print create endpoint')
assert.match(controller, /\/\{id:\\\\d\+\}\/controlled-print\/records/, 'missing controlled print records endpoint')
assert.match(controller, /controlled-print\/print-html/, 'missing controlled print HTML endpoint')

for (const relativePath of [
  'yudao-module-dcc/src/main/java/cn/iocoder/yudao/module/dcc/service/file/DccControlledFilePrintService.java',
  'yudao-module-dcc/src/main/java/cn/iocoder/yudao/module/dcc/service/file/DccControlledFilePrintServiceImpl.java',
  'yudao-module-dcc/src/main/java/cn/iocoder/yudao/module/dcc/dal/dataobject/file/DccControlledFilePrintRecordDO.java',
  'yudao-module-dcc/src/main/java/cn/iocoder/yudao/module/dcc/dal/mysql/file/DccControlledFilePrintRecordMapper.java'
]) {
  assert.equal(fs.existsSync(path.join(backendRoot, relativePath)), true, `missing ${relativePath}`)
}

const errors = readSource('yudao-module-dcc/src/main/java/cn/iocoder/yudao/module/dcc/enums/ErrorCodeConstants.java')
assert.match(errors, /CONTROLLED_FILE_PRINT_NOT_ALLOWED/, 'missing controlled print permission/current error')
assert.match(errors, /CONTROLLED_FILE_PRINT_REQUIRED_FIELD_MISSING/, 'missing controlled print required-field error')

console.log('PASS: DCC controlled print backend contract')
